python/rascaline/rascaline/utils/atomic_density.py

```diff
@@ -0,0 +1,166 @@
+r"""
+.. _python-atomic-density:
+
+Atomic Density
+==============
+
+the atomic density function :math:`g(r)`, often chosen to be a Gaussian or Delta
+function, that defined the type of density under consideration. For a given center atom
+:math:`i` in the structure, the total density function :math:`\rho_i(\boldsymbol{r})`
+around is then defined as :math:`\rho_i(\boldsymbol{r}) = \sum_{j} g(\boldsymbol{r} -
+\boldsymbol{r}_{ij})`.
+
+All atomic densities are based on
+
+.. autoclass:: rascaline.utils.atomic_density.AtomicDensityBase
+    :members:
+    :show-inheritance:
+
+In addition, we provide the following explicit implementations
+
+.. autoclass:: rascaline.utils.atomic_density.DeltaDensity
+    :members:
+    :show-inheritance:
+
+.. autoclass:: rascaline.utils.atomic_density.GaussianDensity
+    :members:
+    :show-inheritance:
+
+.. autoclass:: LodeDensity
+    :members:
+    :show-inheritance:
+
+"""
+from abc import ABC, abstractmethod
+from typing import Union
+
+import numpy as np
+
+
+try:
+    from scipy.special import gamma, gammainc
+
+    HAS_SCIPY = True
+except ImportError:
+    HAS_SCIPY = False
+
+
+class AtomicDensityBase(ABC):
+    """Base class implemententing atomic densities."""
+
+    @abstractmethod
+    def compute(self, positions: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
```

Comment:
We could use `@overload` to do something like

```
def compute(self, positions: float) -> float:

def compute(self, positions: np.ndarray) -> np.ndarray:
```

Reply:
But then we have to provide both functions in all subclasses no?